diff --git a/x.c b/x.c
index d73152b..67a493e 100644
--- a/x.c
+++ b/x.c
@@ -254,6 +254,77 @@ static char *opt_title = NULL;
 
 static uint buttons; /* bit field of pressed buttons */
 
+static int show_welcome_msg = 0;
+
+static char *welcome_cmd[4] = {
+"/bin/sh", "-c", " printf \'\\033[H\\033[2J\';"
+" printf \'  ****************** \\033[1mWelcome to Simple Terminal+ (st+)\\033[0m *******************\n\';"
+" printf \'  ◆ Default configuration file has been generated in: \\033[1m~/.st.conf\\033[0m\n\';"
+" printf \'  ◆ To configure st+ and install nerd-fonts, run: \\033[1m$APPDIR/st-config\\033[0m\n\';"
+" printf \'  ◆ To uninstall st+ and related files, run: \\033[1m$APPDIR/st-config uninstall\\033[0m\n\';"
+" printf \'  ************************************************************************\n\'; exec $SHELL", NULL};
+
+void
+load_config(void)
+{
+	FILE *fp;
+	char path[256];
+	char line[256];
+	char results[128];
+	unsigned int n;
+
+	// Open and read config file
+	const char *home = getenv("HOME");
+	if (!home) return;
+	snprintf(path, sizeof(path), "%s/.st.conf", home);
+	fp = fopen(path, "r");
+
+	// Build config file if missing
+	if (!fp) {
+		show_welcome_msg = 1;
+		fp = fopen(path, "w");
+		if (!fp) return;
+		fprintf(fp, "## General config\n");
+		fprintf(fp, "font=\"%s\"\n", font);
+		fprintf(fp, "rows=\"%d\"\n", rows);
+		fprintf(fp, "cols=\"%d\"\n", cols);
+		fprintf(fp, "alpha=\"%f\"\n", alpha);
+		fprintf(fp, "\n## Primary 16 colors\n");
+		for (n=0; n<16; n++)
+			fprintf(fp, "colorname[%d]=\"%s\"\n", n, colorname[n]);
+		fprintf(fp, "\n## Additional colors for BG, FG, Cursor\n");
+		for (n=256; n<259; n++)
+			fprintf(fp, "colorname[%d]=\"%s\"\n", n, colorname[n]);
+		fclose(fp);
+		return;
+	}
+
+	// Scan configuration values
+	while (fgets(line, sizeof(line), fp)) {
+		if (sscanf(line, "#%s", results) == 1) {} //Skip comments
+		else if (sscanf(line, "font=\"%127[^\"]\"", results) == 1) {
+			font = xmalloc(strlen(results) + 1);
+			strcpy(font, results);
+		}
+		else if (sscanf(line, "rows=\"%5[^\"]\"", results) == 1)
+			rows = strtol(results, NULL, 10);
+		else if (sscanf(line, "cols=\"%5[^\"]\"", results) == 1)
+			cols = strtol(results, NULL, 10);
+		else if (sscanf(line, "alpha=\"%5[^\"]\"", results) == 1)
+			alpha = strtof(results, NULL);
+		else if (sscanf(line, "colorname[%d]=\"%7[^\"]\"", &n, results) == 2) {
+			if ((n<16) || BETWEEN(n, 256, 258)) {
+				colorname[n] = xmalloc(7);
+				strncpy(colorname[n], results, 7);
+			}
+		}
+	}
+
+	// Overwrite options and close file
+	fclose(fp);
+	return;
+}
+
 void
 clipcopy(const Arg *dummy)
 {
@@ -994,6 +1065,11 @@ xloadfonts(const char *fontstr, double fontsize)
 	if (!pattern)
 		die("can't open font %s\n", fontstr);
 
+	// Auto-calculate suitable font size when user specifies size as 0
+	if (FcPatternGetDouble(pattern, FC_PIXEL_SIZE, 0, &fontval) == FcResultMatch) {
+		if ((fontval == 0) && (fontsize == 0)) fontsize = 14.0 * ((float)DisplayHeight(xw.dpy, xw.scr)) / 1080.0;
+	}
+
 	if (fontsize > 1) {
 		FcPatternDel(pattern, FC_PIXEL_SIZE);
 		FcPatternDel(pattern, FC_SIZE);
@@ -1946,7 +2022,10 @@ run(void)
 		}
 	} while (ev.type != MapNotify);
 
-	ttyfd = ttynew(opt_line, shell, opt_io, opt_cmd);
+	if (show_welcome_msg)
+		ttyfd = ttynew(opt_line, shell, opt_io, welcome_cmd); //Show welcome message, ignore -e
+	else
+		ttyfd = ttynew(opt_line, shell, opt_io, opt_cmd);
 	cresize(w, h);
 
 	for (timeout = -1, drawing = 0, lastblink = (struct timespec){0};;) {
@@ -2094,6 +2173,7 @@ run:
 	if (!opt_title)
 		opt_title = (opt_line || !opt_cmd) ? "st" : opt_cmd[0];
 
+	load_config(); //Loads values from config file
 	setlocale(LC_CTYPE, "");
 	XSetLocaleModifiers("");
 	cols = MAX(cols, 1);
