diff --git a/config.def.h b/config.def.h
index 2cd740a..ac550dd 100644
--- a/config.def.h
+++ b/config.def.h
@@ -94,34 +94,31 @@ char *termname = "st-256color";
 unsigned int tabspaces = 8;
 
 /* Terminal colors (16 first used in escape sequence) */
-static const char *colorname[] = {
-	/* 8 normal colors */
-	"black",
-	"red3",
-	"green3",
-	"yellow3",
-	"blue2",
-	"magenta3",
-	"cyan3",
-	"gray90",
-
-	/* 8 bright colors */
-	"gray50",
-	"red",
-	"green",
-	"yellow",
-	"#5c5cff",
-	"magenta",
-	"cyan",
-	"white",
+static char *colorname[] = {
+	/* nord */
+	"#3b4252",  /*  0: black    */
+	"#bf616a",  /*  1: red      */
+	"#a3be8c",  /*  2: green    */
+	"#ebcb8b",  /*  3: yellow   */
+	"#81a1c1",  /*  4: blue     */
+	"#b48ead",  /*  5: magenta  */
+	"#88c0d0",  /*  6: cyan     */
+	"#e5e9f0",  /*  7: white    */
+	"#4c566a",  /*  8: brblack  */
+	"#bf616a",  /*  9: brred    */
+	"#a3be8c",  /* 10: brgreen  */
+	"#ebcb8b",  /* 11: bryellow */
+	"#81a1c1",  /* 12: brblue   */
+	"#b48ead",  /* 13: brmagenta*/
+	"#8fbcbb",  /* 14: brcyan   */
+	"#eceff4",  /* 15: brwhite  */
 
 	[255] = 0,
 
 	/* more colors can be added after 255 to use with DefaultXX */
-	"#cccccc",
-	"#555555",
-	"gray90", /* default foreground colour */
-	"black", /* default background colour */
+	"#2e3440",  /* 256: bg     */
+	"#d8dee9",  /* 257: fg     */
+	"#d8dee9",  /* 258: cursor */
 };
 
 
@@ -129,10 +126,10 @@ static const char *colorname[] = {
  * Default colors (colorname index)
  * foreground, background, cursor, reverse cursor
  */
-unsigned int defaultfg = 258;
-unsigned int defaultbg = 259;
-unsigned int defaultcs = 256;
-static unsigned int defaultrcs = 257;
+unsigned int defaultfg = 257;
+unsigned int defaultbg = 256;
+unsigned int defaultcs = 258;
+static unsigned int defaultrcs = 256;
 
 /*
  * Default shape of cursor
