diff --git a/x.c b/x.c
index d73152b..61c55d0 100644
--- a/x.c
+++ b/x.c
@@ -254,6 +254,51 @@ static char *opt_title = NULL;
 
 static uint buttons; /* bit field of pressed buttons */
 
+void
+load_config(void)
+{
+	FILE *fp;
+	char path[256];
+	char line[256];
+	char results[128];
+
+	// Open and read config file
+	const char *home = getenv("HOME");
+	if (!home) return;
+	snprintf(path, sizeof(path), "%s/.st.conf", home);
+	fp = fopen(path, "r");
+
+	// Build config file if missing
+	if (!fp) {
+		fp = fopen(path, "w");
+		if (!fp) return;
+		fprintf(fp, "font=\"%s\"\n", font);
+		fprintf(fp, "alpha=\"%f\"\n", alpha);
+		fprintf(fp, "rows=\"%d\"\n", rows);
+		fprintf(fp, "cols=\"%d\"\n", cols);
+		fclose(fp);
+		return;
+	}
+
+	// Scan configuration values
+	while (fgets(line, sizeof(line), fp)) {
+		if (sscanf(line, "font=\"%127[^\"]\"", results) == 1) {
+			font = malloc(strlen(results) + 1);
+			strcpy(font, results);
+		}
+		if (sscanf(line, "alpha=\"%5[^\"]\"", results) == 1)
+			alpha = strtof(results, NULL);
+		if (sscanf(line, "rows=\"%5[^\"]\"", results) == 1)
+			rows = strtol(results, NULL, 10);
+		if (sscanf(line, "cols=\"%5[^\"]\"", results) == 1)
+			cols = strtol(results, NULL, 10);
+	}
+
+	// Overwrite options and close file
+	fclose(fp);
+	return;
+}
+
 void
 clipcopy(const Arg *dummy)
 {
@@ -2094,6 +2139,7 @@ run:
 	if (!opt_title)
 		opt_title = (opt_line || !opt_cmd) ? "st" : opt_cmd[0];
 
+	load_config(); //Loads values from config file
 	setlocale(LC_CTYPE, "");
 	XSetLocaleModifiers("");
 	cols = MAX(cols, 1);
